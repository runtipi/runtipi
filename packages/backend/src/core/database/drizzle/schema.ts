import { relations } from 'drizzle-orm';
import { boolean, customType, foreignKey, integer, pgEnum, pgTable, serial, text, timestamp, unique, varchar } from 'drizzle-orm/pg-core';

export const appStatusEnum = pgEnum('app_status_enum', [
  'running',
  'stopped',
  'installing',
  'uninstalling',
  'stopping',
  'starting',
  'missing',
  'updating',
  'resetting',
  'restarting',
  'backing_up',
  'restoring',
]);
export const updateStatusEnum = pgEnum('update_status_enum', ['FAILED', 'SUCCESS']);

export const link = pgTable(
  'link',
  {
    id: serial().primaryKey().notNull(),
    title: varchar({ length: 20 }).notNull(),
    url: varchar().notNull(),
    iconUrl: varchar('icon_url'),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    userId: integer('user_id').notNull(),
    description: varchar({ length: 50 }),
  },
  (table) => {
    return {
      fkLinkUserId: foreignKey({
        columns: [table.userId],
        foreignColumns: [user.id],
        name: 'FK_link_user_id',
      }).onDelete('cascade'),
    };
  },
);

const appConfig = customType<{ data: Record<string, unknown>; driverData: string }>({
  dataType() {
    return 'jsonb';
  },
  toDriver(value: Record<string, unknown>): string {
    return JSON.stringify(value);
  },
});

export const app = pgTable(
  'app',
  {
    id: varchar().primaryKey().notNull(),
    status: appStatusEnum().default('stopped').notNull(),
    lastOpened: timestamp({ withTimezone: true, mode: 'string' }).defaultNow(),
    numOpened: integer().default(0).notNull(),
    config: appConfig('config').notNull(),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    version: integer().default(1).notNull(),
    exposed: boolean().default(false).notNull(),
    domain: varchar(),
    isVisibleOnGuestDashboard: boolean('is_visible_on_guest_dashboard').default(false).notNull(),
    openPort: boolean('open_port').default(true).notNull(),
    exposedLocal: boolean('exposed_local').default(true).notNull(),
    appStoreId: integer('app_store_id').references(() => appStore.id),
  },
  (table) => {
    return {
      uq9478629Fc093D229Df09E560Aea: unique('UQ_9478629fc093d229df09e560aea').on(table.id),
    };
  },
);

export const appRelations = relations(app, ({ one }) => ({
  appStore: one(appStore, {
    fields: [app.appStoreId],
    references: [appStore.id],
  }),
}));

export const user = pgTable(
  'user',
  {
    id: serial().primaryKey().notNull(),
    username: varchar().notNull(),
    password: varchar().notNull(),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    operator: boolean().default(false).notNull(),
    totpSecret: text('totp_secret'),
    totpEnabled: boolean('totp_enabled').default(false).notNull(),
    salt: text(),
    locale: varchar().default('en').notNull(),
    hasSeenWelcome: boolean('has_seen_welcome').default(false).notNull(),
  },
  (table) => {
    return {
      uq78A916Df40E02A9Deb1C4B75Edb: unique('UQ_78a916df40e02a9deb1c4b75edb').on(table.username),
    };
  },
);

export const appStore = pgTable('app_store', {
  id: serial().primaryKey().notNull(),
  hash: varchar().notNull().unique(),
  name: varchar().notNull(),
  enabled: boolean().default(true).notNull(),
  url: varchar().notNull(),
  branch: varchar().default('main').notNull(),
  createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
  updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
  deleted: boolean().default(false).notNull(),
});
