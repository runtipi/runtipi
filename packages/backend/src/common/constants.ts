export const APP_DIR = '/app';
export const DATA_DIR = '/data';
export const APP_DATA_DIR = '/app-data';

export const SESSION_COOKIE_NAME = 'tipi.sid';
export const SESSION_COOKIE_MAX_AGE = 1000 * 60 * 60 * 24;

export const ARCHITECTURES = ['arm64', 'amd64'] as const;
export type Architecture = (typeof ARCHITECTURES)[number];

export const LATEST_RELEASE_URL = 'https://api.github.com/repos/runtipi/runtipi/releases/latest';
