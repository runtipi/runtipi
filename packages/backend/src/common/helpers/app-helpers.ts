export const extractAppId = (id: string) => {
  const separatorIndex = id.indexOf('_');
  if (separatorIndex === -1) {
    throw new Error(`Invalid namespaced app id: ${id}`);
  }
  const appId = id.substring(0, separatorIndex);
  const storeId = id.substring(separatorIndex + 1);

  if (!storeId || !appId) {
    throw new Error(`Invalid namespaced app id: ${id}`);
  }

  return { storeId, appId };
};
