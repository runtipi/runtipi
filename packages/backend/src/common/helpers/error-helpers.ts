import type { ErrorEvent, EventHint } from '@sentry/core';
import cloneDeep from 'lodash.clonedeep';
import validator from 'validator';

const IgnoreErrors = [
  // Innocuous browser errors
  /ResizeObserver loop limit exceeded/,
  /ResizeObserver loop completed with undelivered notifications/,
  // Dark reader extension
  /WeakMap key undefined must be an object or an unregistered symbol/,
  // Docker-compose error
  /no space left on device/,
  /port is already allocated/,
  /address already in use/,
  /Error with your custom app/,
  /cannot assign requested address/,
];

const cleanseUrl = (url: string) => {
  if (validator.isURL(url)) {
    const { pathname, search } = new URL(url);

    return `${pathname}${search}`;
  }

  return url;
};

const shouldIgnoreException = (s: string) => {
  return s && IgnoreErrors.find((pattern) => pattern.test(s));
};

export const cleanseErrorData = (event: ErrorEvent, hint: EventHint) => {
  const result = cloneDeep(event);

  const error = hint && (hint.originalException as Error);

  if (result.transaction) {
    result.transaction = cleanseUrl(result.transaction);
  }

  if (result.breadcrumbs) {
    for (const breadcrumb of result.breadcrumbs) {
      if (breadcrumb.data?.url) {
        breadcrumb.data.url = cleanseUrl(breadcrumb.data.url);
      }
    }
  }

  if (result.exception?.values) {
    for (const exception of result.exception.values) {
      if (exception.stacktrace?.frames) {
        for (const frame of exception.stacktrace.frames) {
          if (frame.filename) {
            frame.filename = cleanseUrl(frame.filename);
          }
        }
      }
    }
  }

  if (error?.message && shouldIgnoreException(error.message)) {
    return null;
  }

  // IF error message starts with 'Command failed: docker-compose' then grab only the 200 last characters
  if (error?.message?.startsWith('Command failed: docker-compose')) {
    // Command failed: docker-compose --env-file /storage/app-data/<app-name>/app.env
    const appName = error.message.split('/')[3];
    const message = error.message.slice(-200);
    result.message = `Error with ${appName}: ${message}`;
  }

  if (result.request?.url) {
    result.request.url = cleanseUrl(result.request.url);
  }

  return result;
};
