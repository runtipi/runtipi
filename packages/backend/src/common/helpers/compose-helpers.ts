import type { Architecture } from '@/common/constants.js';
import { type DynamicCompose, type ServiceInput } from '@runtipi/common/schemas';
import deepmerge from 'deepmerge';

/**
 * Merges architecture-specific overrides with the base configuration
 * @param services The base services from configuration
 * @param overrides The architecture-specific overrides from configuration
 * @param architecture The current system architecture
 * @returns The merged services configuration
 */
export const mergeArchitectureOverrides = (
  services: ServiceInput[],
  overrides: DynamicCompose['overrides'] | undefined,
  architecture: Architecture,
): ServiceInput[] => {
  if (!overrides?.length) {
    return services;
  }

  const architectureOverrides = overrides.find((override) => override.architecture === architecture);

  if (!architectureOverrides) {
    return services;
  }

  const servicesMap = new Map<string, ServiceInput>();
  for (const service of services) {
    servicesMap.set(service.name, service);
  }

  for (const override of architectureOverrides.services) {
    if (!override.name) continue;

    const baseService = servicesMap.get(override.name);
    if (baseService) {
      // Special case for arrays that should be replaced rather than merged
      const arrayMode = {
        arrayMerge: (_target: unknown[], source: unknown[]) => source,
      };

      servicesMap.set(override.name, deepmerge(baseService, override, arrayMode) as ServiceInput);
    }
  }

  return Array.from(servicesMap.values());
};
