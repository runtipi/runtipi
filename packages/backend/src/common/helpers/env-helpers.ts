import crypto from 'node:crypto';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { settingsSchema } from '@/app.dto';
import { EnvUtils } from '@/modules/env/env.utils';
import dotenv from 'dotenv';
import { DATA_DIR } from '../constants';

const OLD_DEFAULT_REPO_URL = 'https://github.com/meienberger/runtipi-appstore';
export const DEFAULT_REPO_URL = 'https://github.com/runtipi/runtipi-appstore';

/**
 * Generates a random seed if it does not exist yet
 */
const generateSeed = async () => {
  const seedFilePath = path.join(DATA_DIR, 'state', 'seed');
  if (!fs.existsSync(seedFilePath)) {
    const randomBytes = crypto.randomBytes(32);
    const seed = randomBytes.toString('hex');
    await fs.promises.writeFile(seedFilePath, seed);
  }
};

/**
 * Returns the architecture of the current system
 */
const getArchitecture = () => {
  const arch = os.arch();

  if (arch === 'arm64') return 'arm64';
  if (arch === 'x64') return 'amd64';

  throw new Error(`Unsupported architecture: ${arch}`);
};

export const generateSystemEnvFile = async (): Promise<Map<string, string>> => {
  const envUtils = new EnvUtils();

  await fs.promises.mkdir(path.join(DATA_DIR, 'state'), { recursive: true });

  const settingsFilePath = path.join(DATA_DIR, 'state', 'settings.json');
  const envFilePath = path.join(DATA_DIR, '.env');

  if (!fs.existsSync(envFilePath)) {
    await fs.promises.writeFile(envFilePath, '');
  }

  const envFile = await fs.promises.readFile(envFilePath, 'utf-8');

  const envMap: Map<string, string> = envUtils.envStringToMap(envFile);
  envMap.set('NODE_ENV', process.env.NODE_ENV || 'production');

  if (!fs.existsSync(settingsFilePath)) {
    await fs.promises.writeFile(settingsFilePath, JSON.stringify({}));
  }

  const settingsFile = await fs.promises.readFile(settingsFilePath, 'utf-8');

  const settings = settingsSchema.partial().safeParse(JSON.parse(settingsFile));

  if (!settings.success) {
    throw new Error(`Invalid settings.json file: ${settings.error.message}`);
  }

  await generateSeed();

  const { data } = settings;

  if (data.appsRepoUrl === OLD_DEFAULT_REPO_URL) {
    data.appsRepoUrl = DEFAULT_REPO_URL;
  }

  const jwtSecret = envMap.get('JWT_SECRET') || envUtils.deriveEntropy('jwt_secret');

  const repoUrl = data.appsRepoUrl || envMap.get('APPS_REPO_URL') || DEFAULT_REPO_URL;
  const hash = crypto.createHash('sha256');
  hash.update(repoUrl);
  const repoId = hash.digest('hex');

  const rootFolderHost = envMap.get('ROOT_FOLDER_HOST') ?? process.env.ROOT_FOLDER_HOST;
  const internalIp = envMap.get('INTERNAL_IP') ?? '127.0.0.1';

  if (!rootFolderHost) {
    throw new Error(
      'Failed to determine root folder host. If you are not running via the CLI, please set the ROOT_FOLDER_HOST environment variable.',
    );
  }

  // Ensure that the app data path does not contain the /app-data suffix
  const fixedAppDataPath = data.appDataPath?.split('/app-data')[0];

  envMap.set('ROOT_FOLDER_HOST', rootFolderHost);
  envMap.set('APPS_REPO_ID', repoId);
  envMap.set('APPS_REPO_URL', data.appsRepoUrl || envMap.get('APPS_REPO_URL') || DEFAULT_REPO_URL);
  envMap.set('TZ', data.timeZone || Intl.DateTimeFormat().resolvedOptions().timeZone);
  envMap.set('INTERNAL_IP', data.listenIp || internalIp);
  envMap.set('DNS_IP', data.dnsIp || envMap.get('DNS_IP') || '9.9.9.9');
  envMap.set('ARCHITECTURE', getArchitecture());
  envMap.set('JWT_SECRET', jwtSecret);
  envMap.set('DOMAIN', data.domain || envMap.get('DOMAIN') || 'example.com');
  envMap.set('RUNTIPI_APP_DATA_PATH', fixedAppDataPath || envMap.get('RUNTIPI_APP_DATA_PATH') || rootFolderHost);
  envMap.set('POSTGRES_HOST', 'runtipi-db');
  envMap.set('POSTGRES_DBNAME', 'tipi');
  envMap.set('POSTGRES_USERNAME', 'tipi');
  envMap.set('POSTGRES_PORT', String(5432));
  envMap.set('DEMO_MODE', typeof data.demoMode === 'boolean' ? String(data.demoMode) : envMap.get('DEMO_MODE') || 'false');
  envMap.set('GUEST_DASHBOARD', typeof data.guestDashboard === 'boolean' ? String(data.guestDashboard) : envMap.get('GUEST_DASHBOARD') || 'false');
  envMap.set('LOCAL_DOMAIN', data.localDomain || envMap.get('LOCAL_DOMAIN') || 'tipi.lan');
  envMap.set(
    'ALLOW_AUTO_THEMES',
    typeof data.allowAutoThemes === 'boolean' ? String(data.allowAutoThemes) : envMap.get('ALLOW_AUTO_THEMES') || 'true',
  );
  envMap.set(
    'ALLOW_ERROR_MONITORING',
    typeof data.allowErrorMonitoring === 'boolean' ? String(data.allowErrorMonitoring) : envMap.get('ALLOW_ERROR_MONITORING') || 'false',
  );
  envMap.set(
    'PERSIST_TRAEFIK_CONFIG',
    typeof data.persistTraefikConfig === 'boolean' ? String(data.persistTraefikConfig) : envMap.get('PERSIST_TRAEFIK_CONFIG') || 'false',
  );

  await fs.promises.writeFile(envFilePath, envUtils.envMapToString(envMap));

  dotenv.config({ path: envFilePath, override: true });

  return envMap;
};
