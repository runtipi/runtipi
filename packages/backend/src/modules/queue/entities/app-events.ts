import { appFormSchema } from '@/modules/app-lifecycle/dto/app-lifecycle.dto';
import type { AppUrn } from '@/types/app/app.types';
import { Injectable } from '@nestjs/common';
import { type ZodStringDef, z } from 'zod';
import { Queue } from '../queue.entity';

const commonAppCommandSchema = z.object({
  command: z.union([
    z.literal('start'),
    z.literal('stop'),
    z.literal('install'),
    z.literal('uninstall'),
    z.literal('reset'),
    z.literal('restart'),
    z.literal('generate_env'),
    z.literal('backup'),
  ]),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  skipEnv: z.boolean().optional().default(false),
  form: appFormSchema,
});

const restoreAppCommandSchema = z.object({
  command: z.literal('restore'),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  filename: z.string(),
  form: appFormSchema,
});

const updateAppCommandSchema = z.object({
  command: z.literal('update'),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  form: appFormSchema,
  performBackup: z.boolean().optional().default(true),
});

export const appEventSchema = commonAppCommandSchema.or(restoreAppCommandSchema).or(updateAppCommandSchema);

export const appEventResultSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type AppEventFormInput = z.input<typeof commonAppCommandSchema>['form'];
export type AppEventForm = z.output<typeof commonAppCommandSchema>['form'];

@Injectable()
export class AppEventsQueue extends Queue<typeof appEventSchema, typeof appEventResultSchema> {}
