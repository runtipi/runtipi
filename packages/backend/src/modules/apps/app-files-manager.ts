import path from 'node:path';
import { extractAppId } from '@/common/helpers/app-helpers';
import { execAsync } from '@/common/helpers/exec-helpers';
import { ConfigurationService } from '@/core/config/configuration.service';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import { LoggerService } from '@/core/logger/logger.service';
import { Injectable } from '@nestjs/common';
import { appInfoSchema } from '../marketplace/dto/marketplace.dto';

@Injectable()
export class AppFilesManager {
  constructor(
    private readonly configuration: ConfigurationService,
    private readonly filesystem: FilesystemService,
    private readonly logger: LoggerService,
  ) {}

  private getInstalledAppsFolder() {
    const { directories } = this.configuration.getConfig();

    return path.join(directories.dataDir, 'apps');
  }

  public getAppPaths(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    return {
      appDataDir: path.join(directories.appDataDir, storeId, appId),
      appInstalledDir: path.join(this.getInstalledAppsFolder(), storeId, appId),
    };
  }

  /**
   * Get the app info from the installed apps apps
   * @param id - The app id
   */
  public async getInstalledAppInfo(id: string) {
    try {
      const { appInstalledDir } = this.getAppPaths(id);

      if (await this.filesystem.pathExists(path.join(appInstalledDir, 'config.json'))) {
        const configFile = await this.filesystem.readTextFile(path.join(appInstalledDir, 'config.json'));
        const parsedConfig = appInfoSchema.safeParse(JSON.parse(configFile ?? ''));

        if (!parsedConfig.success) {
          this.logger.debug(`App ${id} config error:`);
          this.logger.debug(parsedConfig.error);
        }

        if (parsedConfig.success && parsedConfig.data.available) {
          const description = (await this.filesystem.readTextFile(path.join(appInstalledDir, 'metadata', 'description.md'))) ?? '';

          return { ...parsedConfig.data, id, description };
        }
      }
    } catch (error) {
      return null;
    }
  }

  /**
   * Get the docker-compose.json file content from the installed app
   * @param id - The app id
   * @returns The content of docker-compose.yml as a string, or null if not found
   */
  public async getDockerComposeYaml(id: string) {
    const arch = this.configuration.get('architecture');
    const { appInstalledDir } = this.getAppPaths(id);
    let dockerComposePath = path.join(appInstalledDir, 'docker-compose.yml');

    if (arch === 'arm64' && (await this.filesystem.pathExists(path.join(appInstalledDir, 'docker-compose.arm64.yml')))) {
      dockerComposePath = path.join(appInstalledDir, 'docker-compose.arm64.yml');
    }

    let content = null;
    try {
      if (await this.filesystem.pathExists(dockerComposePath)) {
        content = await this.filesystem.readTextFile(dockerComposePath);
      }
    } catch (error) {
      this.logger.error(`Error getting docker-compose.yml for installed app ${id}: ${error}`);
    }

    return { path: dockerComposePath, content };
  }

  /**
   * Get the docker-compose.json file content from the installed app
   * @param id - The app id
   * @returns The content of docker-compose.json as a string, or null if not found
   */
  public async getDockerComposeJson(id: string) {
    const { appInstalledDir } = this.getAppPaths(id);
    const dockerComposePath = path.join(appInstalledDir, 'docker-compose.json');

    let content = null;
    try {
      if (await this.filesystem.pathExists(dockerComposePath)) {
        content = await this.filesystem.readJsonFile(dockerComposePath);
      }
    } catch (error) {
      this.logger.error(`Error getting docker-compose.json for installed app ${id}: ${error}`);
    }

    return { path: dockerComposePath, content };
  }

  /**
   * Write the docker-compose.yml file to the installed app folder
   * @param appId - The app id
   * @param composeFile - The content of the docker-compose.yml file
   */
  public async writeDockerComposeYml(appId: string, composeFile: string) {
    const { appInstalledDir } = this.getAppPaths(appId);
    const dockerComposePath = path.join(appInstalledDir, 'docker-compose.yml');

    await this.filesystem.writeTextFile(dockerComposePath, composeFile);
  }

  public async deleteAppFolder(appId: string) {
    const { appInstalledDir } = this.getAppPaths(appId);
    await this.filesystem.removeDirectory(appInstalledDir);
  }

  public async deleteAppDataDir(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);
    await this.filesystem.removeDirectory(appDataDir);
  }

  public async createAppDataDir(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);
    await this.filesystem.createDirectory(appDataDir);
  }

  /**
   * Set the permissions for the app data directory
   * @param appId - The app id
   */
  public async setAppDataDirPermissions(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);

    await execAsync(`chmod -Rf a+rwx ${appDataDir}`).catch(() => {
      this.logger.error(`Error setting permissions for app ${appId}`);
    });
  }

  public async getAppEnv(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);

    const envPath = path.join(appDataDir, 'app.env');

    let env = '';
    if (await this.filesystem.pathExists(envPath)) {
      env = (await this.filesystem.readTextFile(envPath)) ?? '';
    }

    return { path: envPath, content: env };
  }

  public async writeAppEnv(appId: string, env: string) {
    const { appDataDir } = this.getAppPaths(appId);

    const envPath = path.join(appDataDir, 'app.env');

    await this.filesystem.writeTextFile(envPath, env);
  }

  /**
   * Get the user env file content
   * @param namespacedAppId - The app id
   */
  public async getUserEnv(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    const userEnvFile = path.join(directories.dataDir, 'user-config', storeId, appId, 'app.env');
    let content = null;

    if (await this.filesystem.pathExists(userEnvFile)) {
      content = await this.filesystem.readTextFile(userEnvFile);
    }

    return { path: userEnvFile, content };
  }

  /**
   * Get the user compose file content
   * @param namespacedAppId - The app id
   */
  public async getUserComposeFile(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    const userComposeFile = path.join(directories.dataDir, 'user-config', storeId, appId, 'docker-compose.yml');
    let content = null;

    if (await this.filesystem.pathExists(userComposeFile)) {
      content = await this.filesystem.readTextFile(userComposeFile);
    }

    return { path: userComposeFile, content };
  }
}
