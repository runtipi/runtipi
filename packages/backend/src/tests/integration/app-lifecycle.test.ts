import fs from 'node:fs';
import { APP_DATA_DIR, APP_DIR, DATA_DIR } from '@/common/constants';
import { ConfigurationService } from '@/core/config/configuration.service';
import { DatabaseService } from '@/core/database/database.service';
import { appStore } from '@/core/database/drizzle/schema';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import type { LoggerService } from '@/core/logger/logger.service';
import { AppLifecycleCommandFactory } from '@/modules/app-lifecycle/app-lifecycle-command.factory';
import { AppLifecycleService } from '@/modules/app-lifecycle/app-lifecycle.service';
import { AppStoreRepository } from '@/modules/app-stores/app-store.repository';
import { AppStoreService } from '@/modules/app-stores/app-store.service';
import { AppFilesManager } from '@/modules/apps/app-files-manager';
import { AppHelpers } from '@/modules/apps/app.helpers';
import { AppsRepository } from '@/modules/apps/apps.repository';
import { EnvUtils } from '@/modules/env/env.utils';
import { MarketplaceService } from '@/modules/marketplace/marketplace.service';
import { AppEventsQueue, appEventResultSchema, appEventSchema } from '@/modules/queue/entities/app-events';
import { QueueFactory } from '@/modules/queue/queue.factory';
import { Test } from '@nestjs/testing';
import { fromPartial } from '@total-typescript/shoehorn';
import { beforeAll, beforeEach, describe, expect, it } from 'vitest';
import { mock } from 'vitest-mock-extended';
import waitFor from 'wait-for-expect';
import type { FsMock } from '../__mocks__/fs';
import { createAppInStore } from '../utils/create-app-in-store';
import { type TestDatabase, cleanTestData, createTestDatabase } from '../utils/create-test-database';

let db: TestDatabase;
const DB_NAME = 'applifecycletest';

describe('App lifecycle', () => {
  let appLifecycleService: AppLifecycleService;
  let marketplaceService: MarketplaceService;
  let appsRepository: AppsRepository;
  let configurationService = mock<ConfigurationService>();
  let databaseService = mock<DatabaseService>();
  const loggerService = mock<LoggerService>();
  const queueFactory = new QueueFactory(loggerService);
  const appEventsQueue = queueFactory.createQueue({
    queueName: 'app-events-queue',
    workers: 1,
    eventSchema: appEventSchema,
    resultSchema: appEventResultSchema,
  });

  beforeAll(async () => {
    db = await createTestDatabase(DB_NAME);
  });

  beforeEach(async () => {
    await cleanTestData(db);

    const moduleRef = await Test.createTestingModule({
      providers: [
        AppLifecycleService,
        MarketplaceService,
        AppStoreService,
        AppStoreRepository,
        FilesystemService,
        QueueFactory,
        AppLifecycleCommandFactory,
        AppFilesManager,
        AppsRepository,
        EnvUtils,
        AppHelpers,
        {
          provide: DatabaseService,
          useValue: databaseService,
        },
        {
          provide: AppEventsQueue,
          useValue: appEventsQueue,
        },
      ],
    })
      .useMocker(mock)
      .compile();

    appLifecycleService = moduleRef.get(AppLifecycleService);
    configurationService = moduleRef.get(ConfigurationService);
    databaseService = moduleRef.get(DatabaseService);
    marketplaceService = moduleRef.get(MarketplaceService);
    appsRepository = moduleRef.get(AppsRepository);

    databaseService.db = db;

    configurationService.getConfig.mockReturnValue(
      fromPartial({
        demoMode: false,
        directories: { dataDir: DATA_DIR, appDir: APP_DIR, appDataDir: APP_DATA_DIR },
        internalIp: '127.0.0.1',
        envFilePath: '/data/.env',
        rootFolderHost: '/opt/runtipi',
        userSettings: {
          appDataPath: '/opt/runtipi',
        },
      }),
    );

    await db.insert(appStore).values({ id: 1, url: 'https://appstore.example.com', hash: 'test', name: 'test', enabled: true }).execute();
    await marketplaceService.initialize();
  });

  describe('install app', () => {
    it('should successfully install app and create expected directory structure', async () => {
      // arrange
      const appInfo = await createAppInStore(1, { id: 'test' });
      const appId = `${appInfo.id}_1`;

      // act
      await appLifecycleService.installApp({ appId, form: {} });

      await waitFor(async () => {
        const app = await appsRepository.getApp(appId);
        expect(app?.status).toBe('running');
      });

      // assert
      expect((fs as unknown as FsMock).tree()).toMatchSnapshot();
      const yml = await fs.promises.readFile(`${DATA_DIR}/apps/1/${appInfo.id}/docker-compose.yml`, 'utf-8');
      expect(yml).toMatchSnapshot();
    });

    it('should not delete an existing app-data folder even if the app is reinstalled', async () => {
      // arrange
      const appInfo = await createAppInStore(1, { id: 'test2' });
      const appId = `${appInfo.id}_1`;

      await fs.promises.mkdir(`${APP_DATA_DIR}/1_test2/data`, { recursive: true });
      await fs.promises.writeFile(`${APP_DATA_DIR}/1_test2/data/test.txt`, 'test');

      await appLifecycleService.installApp({ appId, form: {} });

      await waitFor(async () => {
        const app = await appsRepository.getApp(appId);
        expect(app?.status).toBe('running');
      });

      // assert
      expect((fs as unknown as FsMock).tree()).toMatchSnapshot();
    });
  });
});
