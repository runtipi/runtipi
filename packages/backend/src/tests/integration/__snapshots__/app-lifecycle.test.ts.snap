// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App lifecycle > install app > should not delete an existing app-data folder even if the app is reinstalled 1`] = `
"/
├─ app-data/
│  ├─ 1/
│  │  └─ test2/
│  │     ├─ app.env
│  │     └─ data/
│  └─ 1_test2/
│     └─ data/
│        └─ test.txt
└─ data/
   ├─ .env
   ├─ apps/
   │  └─ 1/
   │     └─ test2/
   │        ├─ config.json
   │        ├─ data/
   │        ├─ docker-compose.json
   │        └─ docker-compose.yml
   ├─ backups/
   ├─ repos/
   │  └─ 1/
   │     └─ apps/
   │        └─ test2/
   │           ├─ config.json
   │           ├─ data/
   │           └─ docker-compose.json
   └─ state/
      └─ seed"
`;

exports[`App lifecycle > install app > should successfully install app and create expected directory structure 1`] = `
"/
├─ app-data/
│  └─ 1/
│     └─ test/
│        ├─ app.env
│        └─ data/
└─ data/
   ├─ .env
   ├─ apps/
   │  └─ 1/
   │     └─ test/
   │        ├─ config.json
   │        ├─ data/
   │        ├─ docker-compose.json
   │        └─ docker-compose.yml
   ├─ backups/
   ├─ repos/
   │  └─ 1/
   │     └─ apps/
   │        └─ test/
   │           ├─ config.json
   │           ├─ data/
   │           └─ docker-compose.json
   └─ state/
      └─ seed"
`;

exports[`App lifecycle > install app > should successfully install app and create expected directory structure 2`] = `
"services:
  test_1:
    image: test
    container_name: test_1
    restart: unless-stopped
    networks:
      - tipi_main_network
    environment:
      TEST: test
    ports:
      - \${APP_PORT}:80
    labels:
      generated: true
      traefik.enable: false
      traefik.http.middlewares.test-web-redirect.redirectscheme.scheme: https
      traefik.http.services.test.loadbalancer.server.port: "80"
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true

"
`;
